agent/src/bpf/audit.bpf.c

```diff
@@ -226,6 +234,126 @@ record_blob_data (struct pt_regs *ctx, long context, const char *key_ptr)
   return err;
 }
 
+/* This is similar to record_blob_data but always assumes VALUE_PTR is
+ * present.
+ */
+static __always_inline int
+record_blob_data_explicit (struct pt_regs *ctx, long context, const char *key_ptr,
+			   const void *value_ptr, long value_size)
+{
+  int err;
+
+  struct audit_blob_data_event_st *event =
+    bpf_ringbuf_reserve (&ringbuf,
+			 sizeof(struct audit_blob_data_event_st),
+			 0);
+  if (!event)
+    {
+      DEBUG ("unable to allocate ringbuf entry: %ld\n", -ENOMEM);
+      return -ENOMEM;
+    }
+
+  populate_event_header (&event->base.header,
+			 AUDIT_EVENT_DATA,
+			 sizeof(*event),
+			 context);
+
+  event->base.type = AUDIT_DATA_BLOB;
+  err = bpf_probe_read_user_str (event->base.key, KEY_SIZE, (void *)key_ptr);
+  if (err < 0)
+    {
+      DEBUG ("unable to read event key: %ld\n", err);
+      goto error;
+    }
+
+  if (value_size > 0)
+    {
+      value_size &= (VALUE_SIZE - 1);
```

Comment:
Not a fan of this construct. Can we go unsigned instead? Skip processing all the invalid values just so that we don't end up supporting that later?

Reply:
Afaik this is the only way to tell the eBPF verifier that a value is in the range.

Reply:
It seems to be fine with, say, `value_size = value_size > VALUE_SIZE - 1 ? VALUE_SIZE - 1 : value_size;` Much longer unless macroed, but works for a wider range of VALUE_SIZEs.

Reply:
If I rewrite that way, I get:
```console
93: (6d) if r1 s> r2 goto pc+1 95: R0=0 R1=63 R2=scalar(smax=62) R6=ringbuf_mem(ref_obj_id=12,off=0,imm=0) R7=0 R8=ctx(off=0,imm=0) R9=ringbuf_mem(ref_obj_id=12,off=0,imm=0) R10=fp0 fp-8=mmmmmmmm fp-16=mmmmmmmm fp-24=mmmmmmmm refs=12
; value_size = value_size > VALUE_SIZE - 1 ? VALUE_SIZE - 1 : value_size;
95: (7b) *(u64 *)(r10 -8) = r2        ; R2=scalar(smax=62) R10=fp0 fp-8_w= refs=12
; err = bpf_probe_read_user (event->value, value_size, (void *)value_ptr);
96: (79) r3 = *(u64 *)(r10 -16)       ; R3_w=scalar() R10=fp0 fp-16=mmmmmmmm refs=12
; err = bpf_probe_read_user (event->value, value_size, (void *)value_ptr);
97: (07) r9 += 112                    ; R9_w=ringbuf_mem(ref_obj_id=12,off=112,imm=0) refs=12
; err = bpf_probe_read_user (event->value, value_size, (void *)value_ptr);
98: (bf) r1 = r9                      ; R1_w=ringbuf_mem(ref_obj_id=12,off=112,imm=0) R9_w=ringbuf_mem(ref_obj_id=12,off=112,imm=0) refs=12
99: (85) call bpf_probe_read_user#112
R2 min value is negative, either use unsigned or 'var &= const'
```
We had a similar discussion [before](https://github.com/latchset/crypto-auditing/pull/40#discussion_r1345585689). Let's treat this as an idiom.

Reply:
OK

Reply:
Actually, it might be better using unsigned value, as systemtap probes have some [distinction](https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation#Argument_Format) between signed/unsigned integer types. If that works, I would remove this `&=` and use two's complement for the special values of `value_size`, i.e., `(unsigned long)-2 == 0xfffffffffffffffe` instead of `-2`.

Reply:
I had to keep `array_size &= MAX_EVENTS - 1;` but other uses have been removed.